packages/presentation/presentation_publishing/interfaces/fetch/on_fetch_context_changed.ts

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { debounce } from 'lodash';
+import { combineLatest, Observable, skip, Subscription } from 'rxjs';
+import { AggregateQuery, Filter, Query, TimeRange } from '@kbn/es-query';
+import {
+  apiPublishesTimeRange,
+  apiPublishesUnifiedSearch,
+  PublishesTimeRange,
+  PublishesUnifiedSearch,
+} from './publishes_unified_search';
+import { apiPublishesSearchSession, PublishesSearchSession } from './publishes_search_session';
+import { apiHasParentApi, HasParentApi } from '../has_parent_api';
+import { apiPublishesReload } from './publishes_reload';
+
+export interface FetchContext {
+  isReload: boolean;
+  filters: Filter[] | undefined;
+  query: Query | AggregateQuery | undefined;
+  searchSessionId: string | undefined;
+  timeRange: TimeRange | undefined;
+  timeslice: [number, number] | undefined;
+}
+
+function getFetchContext(api: unknown, isReload: boolean) {
+  const typeApi = api as Partial<
```

Comment:
This pattern of casting to `Partial<T>` then using optional chaining when dealing with the contents is roughly as type safe as a type guard and is pretty straightforward. I also think its useful enough to be used in a couple other places. Maybe in the Presentation panel for instance, where nothing is required but there are plenty of things that _could_ exist.

Food for thought though, if we start using this pattern more, should we provide some sort of helper type that prevents folks from casting like this _without_ using the partial? Because that could result in runtime issues pretty easily.